rend3-routine/src/skinning/mod.rs

```diff
@@ -0,0 +1,272 @@
+                // vector lengths.
+                joint_matrices_ptr
+                    .add(joint_matrix_idx as usize)
+                    .write_unaligned(joint_matrix.to_cols_array_2d());
+                joint_matrix_idx += 1;
+            }
+        }
+    }
+
+    drop(skinning_input_data);
+    drop(joint_matrices_data);
+    gpu_skinning_inputs.unmap();
+    joint_matrices.unmap();
+
+    PreSkinningBuffers {
+        gpu_skinning_inputs,
+        joint_matrices,
+    }
+}
+
+pub struct GpuSkinner {
+    pub pipeline: ComputePipeline,
+    pub vertex_buffers_bgl: BindGroupLayout,
+    pub joint_matrices_bgl: BindGroupLayout,
+    pub skinning_inputs_bgl: BindGroupLayout,
+}
+
+impl GpuSkinner {
+    const WORKGROUP_SIZE: u32 = 64;
+
+    pub fn new(device: &wgpu::Device) -> GpuSkinner {
+        let storage_buffer_ty = |read_only| wgpu::BindingType::Buffer {
+            ty: wgpu::BufferBindingType::Storage { read_only },
+            has_dynamic_offset: false,
+            min_binding_size: None,
+        };
+
+        // Bind group 0 contains some vertex buffers bound as storage buffers
+        let vertex_buffers_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Positions
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Normals
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Tangents
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Joint indices
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Joint weights
+            bglb.build(device, Some("Gpu skinning vertex buffers"))
+        };
+
+        // Bind group 1 contains the joint matrices global buffer. Each
+        // skinning input struct contains an offset into this array.
+        let joint_matrices_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(true), None);
+            bglb.build(device, Some("Gpu skinning joint matrices"))
+        };
+
+        // Bind group 2 contains the pre skinning inputs. This uses dynamic
+        // offsets because there is one dispatch per input, and the offset is
+        // used to indicate which is the current input to the shader.
+        //
+        // NOTE: This would be an ideal use case for push constants, but they are
+        // not available on all platforms so we need to use this workaround.
+        let skinning_inputs_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(
+                wgpu::ShaderStages::COMPUTE,
+                wgpu::BindingType::Buffer {
+                    ty: wgpu::BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: None,
```

Comment:
We should set min binding sizes as this removes draw time work for wgpu

Reply:
I believe I did that right, using the size of the `GpuSkinningInput` for the uniform buffer, and the size of a single element for the vertex buffers. But please double check just in case